### PR TITLE
feat(DLV-534): added support to send promotion result feedback via co…

### PR DESCRIPTION
**Pull request by pcanilho**
…mmit status

### CODE REVIEW

**Comment by pcanilho** [comment]
- [ ] Pending validation

File: cmd/root_args.go [hunk 1]
```diff
@@ -51,6 +51,12 @@ var envMapString = map[*string]boundEnvVar[string]{
 		Env:         helpers.Ptr("DYNAMIC_PROMOTION_KEY"),
 		Default:     helpers.Ptr("gitops-promotion-path"),
 	},
+	&feedbackCommitStatusContext: {
+		Name:        "feedback-commit-status-context",
+		Description: "The context key to use when pushing the commit status to the repository. Supported placeholders: {source}, {target}",
+		Env:         helpers.Ptr("FEEDBACK_COMMIT_STATUS_CONTEXT"),
+		Default:     helpers.Ptr("gitops/promotion :: {source} → {target}"),
```
Comment (isometry) [inline cmd/root_args.go:58]: Given that we're constrained in the number of characters that GitHub will display before it truncates the string, I suggest we minimise the default (and keep it easy to type without having to revert to UTF-8):
```suggestion
		Default:     helpers.Ptr("{source}/{target}"),
```

File: cmd/root_args.go [hunk 2]
```diff
@@ -70,6 +76,18 @@ var envMapBool = map[*bool]boundEnvVar[bool]{
...
 	},
+	&feedbackCommitStatus: {
+		Name:        "feedback-commit-status",
+		Description: "Enable feedback commit status",
+		Env:         helpers.Ptr("FEEDBACK_COMMIT_STATUS"),
+		Default:     helpers.Ptr(true),
+	},
+	&fetchRateLimits: {
+		Name:        "fetch-rate-limits",
+		Description: "Enable per-event fetching of rate limits and corresponding logs decoration",
```
Comment (isometry) [inline cmd/root_args.go:87]: Given that _every_ interaction with the GitHub API returns current usage in the response headers, I feel we're just introducing potential issues (i.e. hitting secondary rate limits) by directly polling the rate limit endpoint…

File: internal/controllers/github.go [hunk 3]
```diff
@@ -284,6 +284,110 @@ func (g *GitHub) FastForwardRefToSha(pCtx *promotion.Context) error {
...
+	feedbackLogger := pCtx.Logger.WithGroup("feedback:commit-status")
+	// Skip if the context is missing required fields
+	if pCtx.Owner == nil || pCtx.Repository == nil || pCtx.HeadSHA == nil {
+		feedbackLogger.Debug("ignoring promotion feedback commit status due to missing context fields",
+			slog.Any("context", pCtx))
+		return nil
+	}
+
+	// Set the target URL if a pull request is present
+	if pCtx.PullRequest == nil {
```
Comment (isometry) [inline internal/controllers/github.go:313]: Comment doesn't appear to correspond to code…

File: internal/controllers/github.go [hunk 4]
```diff
@@ -284,6 +284,110 @@ func (g *GitHub) FastForwardRefToSha(pCtx *promotion.Context) error {
...
+
+	// Infer the commit status from the provided promotion.Result and error
+	stages, index := pCtx.Promoter.Stages, pCtx.Promoter.StageIndex(*pCtx.HeadRef)
+	progress := fmt.Sprintf("%d/%d", index+1, len(stages))
+
+	// Set the commit status message accordingly
+	var msg string
+	switch commitStatus {
+	case CommitStatusSuccess:
+		msg = "✅ Promotion successful :: {progress}"
```
Comment (isometry) [inline internal/controllers/github.go:338]: Let's simplify…
```suggestion
		msg = "✅ {progress} @ {iso8601-timestamp}"
```

File: internal/controllers/github.go [hunk 5]
```diff
@@ -284,6 +284,110 @@ func (g *GitHub) FastForwardRefToSha(pCtx *promotion.Context) error {
...
+	case CommitStatusFailure:
+		msg = fmt.Sprintf("❌ Promotion failed: %v", promotionError)
```
Comment (isometry) [inline internal/controllers/github.go:340]:
```suggestion
		msg = fmt.Sprintf("❌ %v @ {iso8601-timestamp}", promotionError)
```

File: internal/controllers/github.go [hunk 6]
```diff
@@ -284,6 +284,110 @@ func (g *GitHub) FastForwardRefToSha(pCtx *promotion.Context) error {
...
+	case CommitStatusError:
+		msg = fmt.Sprintf("💥 Promotion internal error: %v", promotionError)
```
Comment (isometry) [inline internal/controllers/github.go:342]:
```suggestion
		msg = fmt.Sprintf("💥 %v @ {iso8601-timestamp}", promotionError)
```

File: internal/handler/handler.go [hunk 7]
```diff
@@ -164,222 +192,259 @@ func (h *Handler) Process(body []byte, headers map[string]string) (response help
...
-		return helpers.Response{Body: err.Error(), StatusCode: http.StatusInternalServerError}, err
+		logger.Warn("failed to store event in S3", slog.Any("error", err))
+		return &promotion.Result{
+			Response: helpers.Response{Body: err.Error(), StatusCode: http.StatusInternalServerError},
+		}, nil
 	}
 
-	pCtx := &promotion.Context{
-		EventType:  &eventType,
+	result.Context = &promotion.Context{
```
Comment (isometry) [inline internal/handler/handler.go:215]: Could we not have simplified the refactor here and maintained the cleaner semantic with a simple:
```golang
pCtx := &promotion.Context{
  …
}
result.Context = pCtx
```
…and continued to updated `pCtx` everywhere else?

Comment (pcanilho) [reply internal/handler/handler.go:215]: The reason as to why I did will be made much more apparent in the upcoming prefactor PR. :-)

File: internal/runtime/runtime.go [hunk 8]
```diff
@@ -93,6 +102,30 @@ func (r *Runtime) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
...
+		if statusErr := r.Handler.SendFeedbackCommitStatus(promotionResult, err, status); statusErr != nil {
+			r.logger.Error("failed to send feedback commit status", slog.Any("error", statusErr))
+		}
+	}
+
+	// Rate limits: if fetchRateLimits is enabled, fetch rate limits
+	if rateLimits, err := r.Handler.RateLimits(promotionResult); err != nil {
+		r.logger.Warn("failed to fetch rate limits", slog.Any("error", err))
+	} else {
+		r.logger.Info("rate limits fetched", slog.Any("rateLimits", rateLimits))
```
Comment (isometry) [inline internal/runtime/runtime.go:129]:
```suggestion
		r.logger.Info("GitHub rate limits", slog.Any("rateLimits", rateLimits))
```

File: internal/runtime/runtime.go [hunk 9]
```diff
@@ -93,6 +102,30 @@ func (r *Runtime) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
...
+		if errors.As(err, &promotionErr) {
+			status = controllers.CommitStatusError
+		}
+		if statusErr := r.Handler.SendFeedbackCommitStatus(promotionResult, err, status); statusErr != nil {
+			r.logger.Error("failed to send feedback commit status", slog.Any("error", statusErr))
+		}
+	}
+
+	// Rate limits: if fetchRateLimits is enabled, fetch rate limits
+	if rateLimits, err := r.Handler.RateLimits(promotionResult); err != nil {
```
Comment (isometry) [inline internal/runtime/runtime.go:126]: We need to be smart here: this should only return the rate limits from the most recent API call, and definitely should *not* call `/rate_limits` in the case of invalid requests.

**Review by isometry — CHANGES_REQUESTED** [review]
Comments inline.

**Comment by pcanilho** [comment]
Changes done:
image

image

**Review by isometry — APPROVED** [review]

**Pull request merged**